aiida/orm/nodes/data/singlefile.py

```diff
@@ -23,10 +24,15 @@
 class SinglefileData(Data):
     """Data class that can be used to store a single file in its repository."""
 
-    def __init__(self, filepath, **kwargs):
+    def __init__(self, file, **kwargs):
+        """Construct a new instance and set the contents to that of the file.
+
+        :param file: an absolute filepath or filelike object whose contents to copy
+        """
+        # pylint: disable=redefined-builtin
         super(SinglefileData, self).__init__(**kwargs)
-        if filepath is not None:
-            self.put_object_from_file(filepath)
+        if file is not None:
```

Comment:
so, file is a required constructor argument but we want to allow passing `None`?
shouldn't we just let it fail if someone passes `None`?

Reply:
Agreed, does not make any sense, however, it might be nice to not 100% require a file be passed. For example sub classes may want different behavior. The `CifData` class currently also allows to create it from an `ase` or `values` object, in which case a `file` is not allowed. However, it still needs to call the parent, which will then fail. Leaving aside whether this `CifData` implementation makes sense, the idea that a sub class will set the file differently somehow, is a feasible use-case no?

Reply:
yeah, also upfdata might make problems. let's leave it as it is for the moment.

Reply:
I also just noticed that Github's code highlighting also seems to think that `file` is somehow special despite not being an official reserved keyword. Weird..

Reply:
turns out it's a `builtin`
https://stackoverflow.com/a/24942363/1069467

Reply:
aha, good to know